airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryStagingConsumerFactory.java

```diff
@@ -46,28 +59,76 @@ public AirbyteMessageConsumer create(final JsonNode config,
                                        final BufferCreateFunction onCreateBuffer,
                                        final Function<JsonNode, BigQueryRecordFormatter> recordFormatterCreator,
                                        final Function<String, String> tmpTableNameTransformer,
-                                       final Function<String, String> targetTableNameTransformer) {
+                                       final Function<String, String> targetTableNameTransformer,
+                                       final BigQuerySqlGenerator sqlGenerator,
+                                       final BigQueryDestinationHandler destinationHandler,
+                                       final ParsedCatalog parsedCatalog,
+                                       final String defaultNamespace)
+      throws InterruptedException {
+    boolean use1s1t = TypingAndDedupingFlag.isDestinationV2();
     final Map<AirbyteStreamNameNamespacePair, BigQueryWriteConfig> writeConfigs = createWriteConfigs(
         config,
         catalog,
+        parsedCatalog,
         recordFormatterCreator,
         tmpTableNameTransformer,
         targetTableNameTransformer);
 
+    final var overwriteStreamsWithTmpTable = createFinalTables(use1s1t, parsedCatalog, destinationHandler, sqlGenerator);
+    ThrowingConsumer<AirbyteStreamNameNamespacePair, InterruptedException> typeAndDedupeStreamFunction = typingAndDedupingStreamConsumer(
+        sqlGenerator,
+        destinationHandler,
+        parsedCatalog,
+        use1s1t,
+        overwriteStreamsWithTmpTable);
+
+    ThrowingConsumer<BigQueryWriteConfig, InterruptedException> replaceFinalTableConsumer =
+        getReplaceFinalTableConsumer(
+            use1s1t,
+            sqlGenerator,
+            destinationHandler,
+            overwriteStreamsWithTmpTable,
+            parsedCatalog);
+
     return new BufferedStreamConsumer(
         outputRecordCollector,
         onStartFunction(bigQueryGcsOperations, writeConfigs),
         new SerializedBufferingStrategy(
             onCreateBuffer,
             catalog,
-            flushBufferFunction(bigQueryGcsOperations, writeConfigs, catalog)),
-        onCloseFunction(bigQueryGcsOperations, writeConfigs),
+            flushBufferFunction(bigQueryGcsOperations, writeConfigs, catalog, typeAndDedupeStreamFunction)),
+        onCloseFunction(bigQueryGcsOperations, writeConfigs, replaceFinalTableConsumer),
         catalog,
-        json -> true);
+        json -> true,
+        defaultNamespace);
+  }
+
+  @FunctionalInterface
```

Comment:
Should probably extract this to a better place, not required for merging but we should do it before we forget.

Reply:
there's a CheckedConsumer in airbyte-commons which is defined identically, I'll make the swap.

Reply:
good find!